x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/tool/GenerateElasticPasswordHash.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.tool;
+
+import joptsimple.OptionSet;
+
+import org.elasticsearch.cli.EnvironmentAwareCommand;
+import org.elasticsearch.cli.Terminal;
+import org.elasticsearch.common.settings.KeyStoreWrapper;
+import org.elasticsearch.common.settings.SecureString;
+import org.elasticsearch.env.Environment;
+import org.elasticsearch.xpack.core.XPackSettings;
+import org.elasticsearch.xpack.core.security.authc.support.Hasher;
+
+import static org.elasticsearch.xpack.security.tool.CommandUtils.generatePassword;
+
+public class GenerateElasticPasswordHash extends EnvironmentAwareCommand {
+
+    public static final String KEYSTORE_SETTING_NAME = "autoconfiguration.password_hash";
+
+    public GenerateElasticPasswordHash() {
+        super("Generates a password hash for for the elastic user and stores it in  elasticsearch.keystore");
+    }
+
+    public static void main(String[] args) throws Exception {
+        exit(new GenerateElasticPasswordHash().main(args, Terminal.DEFAULT));
+    }
+
+    /**
+     * This is called by the package installers. It generates a random strong password for the elastic user and stores a salted
+     * hash of it in the elasticsearch keystore , in the autoconfiguration.password_hash setting. This is subsequently picked up
+     * by the node on startup and is set as the password of the elastic user in the security index.
+     */
+    @Override protected void execute(Terminal terminal, OptionSet options, Environment env) throws Exception {
+        final Hasher hasher = Hasher.resolve(XPackSettings.PASSWORD_HASHING_ALGORITHM.get(env.settings()));
```

Comment:
very good point, will do!

Reply:
I had a change of heart while addressing the feedback. I originally had it so as I expected to call ConfigInitialNode from postinst before this. As such the keystore would exist and would be obfuscated. I think it makes sense to retain this order of execution as the security configuration makes sense to be in place even if we fail to set the elastic password hash as this is a recoverable situation ( we print in postinst that they should run the reset tool to get a pwd, instead of printing the pwd out )